internal/manifests/persistentvolume/persistentvolumeclaim.go

```diff
@@ -0,0 +1,103 @@
+package persistentvolume
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/ViaQ/logerr/kverrors"
+	"github.com/ViaQ/logerr/log"
+	"github.com/openshift/elasticsearch-operator/internal/manifests/status"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ComparePVCFunc is the type for functions that compare two persistentvolumeclaims.
+// Return true if two persistentvolumeclaim are equal.
+type ComparePVCFunc func(current, desired *corev1.PersistentVolumeClaim) bool
+
+// MutatePVCFunc is the type for functions that mutate the current persistentvolumeclaim
+// by applying the values from the desired persistentvolumeclaim.
+type MutatePVCFunc func(current, desired *corev1.PersistentVolumeClaim)
+
+// CreateOrUpdatePVC attempts first to create the given persistentvolumeclaim. If the
+// persistentvolumeclaim already exists and the provided comparison func detects any changes
+// an update is attempted. Updates are retried with backoff (See retry.DefaultRetry).
+// Returns the operation result (See status.OperationResultType) and eventually an error.
+func CreateOrUpdatePVC(ctx context.Context, c client.Client, pvc *corev1.PersistentVolumeClaim, cmp ComparePVCFunc, mutate MutatePVCFunc) (status.OperationResultType, error) {
+	err := c.Create(ctx, pvc)
+	if err == nil {
+		return status.OperationResultCreated, nil
+	}
+
+	if !apierrors.IsAlreadyExists(kverrors.Root(err)) {
+		return status.OperationResultNone, kverrors.Wrap(err, "failed to create persistentvolumeclaim",
+			"name", pvc.Name,
+			"namespace", pvc.Namespace,
+		)
+	}
+
+	current := pvc.DeepCopy()
```

Comment:
no need for a deep copy if we're going to get...

Reply:
Done